hledger/Hledger/Cli/Commands/Balance.hs

```diff
@@ -319,12 +323,58 @@ balance opts@CliOpts{reportopts_=ropts} j = do
                 "csv" -> \ropts r -> (++ "\n") $ printCSV $ balanceReportAsCsv ropts r
                 _     -> balanceReportAsText
           writeOutput opts $ render ropts report
-        _ -> do
+          
+        _ | boolopt "budget" rawopts -> do
+          let budget = budgetJournal opts j
+              j' = budgetRollUp opts budget j
+              report = multiBalanceReport ropts (queryFromOpts d ropts) j'
+              budgetReport = multiBalanceReport ropts (queryFromOpts d ropts) budget
+              render = case format of
+                -- XXX: implement csv rendering
+                "csv" -> (++ "\n") . printCSV . multiBalanceReportAsCsv ropts
+                _     -> multiBalanceReportWithBudgetAsText ropts budgetReport
+          writeOutput opts $ render report
+          
+          | otherwise -> do
           let report = multiBalanceReport ropts (queryFromOpts d ropts) j
               render = case format of
-                "csv" -> \ropts r -> (++ "\n") $ printCSV $ multiBalanceReportAsCsv ropts r
-                _     -> multiBalanceReportAsText
-          writeOutput opts $ render ropts report
+                "csv" -> (++ "\n") . printCSV . multiBalanceReportAsCsv ropts
+                _     -> multiBalanceReportAsText ropts
+          writeOutput opts $ render report
+
+-- | Re-map account names to closet parent with periodic transaction from budget.
+-- Accounts that dont have suitable parent are either remapped to "<unbudgeted>:topAccount" 
+-- or left as-is if --show-unbudgeted is provided 
+budgetRollUp :: CliOpts -> Journal -> Journal -> Journal
+budgetRollUp CliOpts{rawopts_=rawopts} budget j = j { jtxns = remapTxn <$> jtxns j }
```

Comment:
You may want to move this to a separate module and use them from `hledger-budget.hs` as well. Eventually we may want to use it for `register` report as well.